### DISCUSSION TITLE
Tool creation tool

### DISCUSSION THREAD

**Discussion by steve-hickman-epistimis** [eclipse-sirius/sirius-desktop#180] [Ideas]
I don't think this exists yet so I'm posting this as an idea:  I would like the ability to define a 'tool creation tool' for end users to use. Currently, specifiers can create new Node Creation tools and user can use only the Node Creation tools that have been pre-specified.  I would like to split the difference and define a Node Creation Tool Template that pre-specifies most of the functionality that new Node Creation Tools would need but leaves a few things modifiable at the user level. This might be preferable to have a separate tool palette from the current one where the Tool Creation Templates are displayed. When a user selects one, they fill in the remaining details and the result is a new Node Creation Tool in the standard tool palette that can then be used to create instances of that Node type. 

Make sense?

**Comment by pcdavid** [comment]
> Make sense?

Functionally, yes, but it's quite different from the way Sirius currently works. Not sure we would want to integrate something like this directly.

It *might* be possible to "hack" something using the existing mechanisms if you are motivated.
Some pointers/ideas:
* The graphical palette is setup in `org.eclipse.sirius.diagram.ui.tools.internal.palette.PaletteManagerImpl.addDefaultTools(DDiagram)`, but it gets its info about which tools are available from the `diagram.getUiState().getToolSections()`. I don't think you want (or even can) change `PaletteManagerImpl`'s behavior (without forking).
* The `UiState.getToolSections()` is computed/refreshed in `org.eclipse.sirius.diagram.tools.api.management.ToolManagement.computeAvailableTools(Session, boolean)`. In particular, in `ToolManagement.getDefaultTools()` we look at 2 EMF resources which are loaded in the session's ResourceSet and contain default/shared tool definitions. One is [`DiagramEnvironment.odesign`](https://github.com/eclipse-sirius/sirius-desktop/blob/master/plugins/org.eclipse.sirius.diagram/model/DiagramEnvironment.odesign). It's normally a static file, but assuming we don't mark it as read-only somewhere *if* you get it at runtime (see how `ToolManagement.getDefaultTools()` [does it](https://github.com/eclipse-sirius/sirius-desktop/blob/master/plugins/org.eclipse.sirius.diagram/src/org/eclipse/sirius/diagram/tools/api/management/ToolManagement.java#L543), then _maybe_ if you add new tool definitions inside of it they will be picked up and appear in the palette. Maybe. This is all speculative.

Even if this works, this leaves the issue of _where_ to actually store the concrete tools definitions in a more pesistent way so that they do not disappear when the session is closed.
